src/wmtk/simplex/tuples_preserving_primitive_types.cpp

```diff
@@ -0,0 +1,31 @@
+#include "tuples_preserving_primitive_types.hpp"
+
+#include <cassert>
+#include <wmtk/Mesh.hpp>
+#include <wmtk/utils/primitive_range.hpp>
+
+namespace wmtk::simplex {
+
+std::vector<Tuple> tuples_preserving_primitive_types(
+    const Mesh& mesh,
+    const Tuple& t,
+    const PrimitiveType simplex_ptype,
+    const PrimitiveType face_ptype)
+{
+    std::vector<PrimitiveType> switch_tuple_types =
+        wmtk::utils::primitive_range(simplex_ptype, face_ptype);
+
+    Tuple t_iter = t;
+
+    std::vector<Tuple> intersection_tuples;
+    do {
+        intersection_tuples.emplace_back(t_iter);
+        for (size_t i = 1; i < switch_tuple_types.size() - 1; ++i) {
+            t_iter = mesh.switch_tuple(t_iter, switch_tuple_types[i]);
+        }
+    } while (t != t_iter);
```

Comment:
does this general function really work via circulation? i think we might ahve discussed how this is implemented in some nice way, but i don't undertand it right now

Reply:
I hope it does. It worked in my few test cases and I cannot think of a case where it does not work but I cannot guarantee that it works all the time.